docs: add language none option to code-example

kapunahelewong opened the pull request:
Adds language none option to code-example so writers can opt out of color in a code snippet.

mary-poppins commented:
You can preview 22e1fe2 at https://pr41335-22e1fe2.ngbuilds.io/.

gkalpak commented on aio/src/app/custom-elements/code/code.component.ts:
```diff
@@ -109,14 +110,16 @@ export class CodeComponent implements OnChanges {
 
   private formatDisplayedCode() {
     const leftAlignedCode = leftAlign(this.code);
+    const skipPrettify = of(undefined);
```
Hm...I just realized that currently we don't support linenums when skipping prettify 😱
I think it is fine for now. Maybe log a warning if `getLinenums()` returns anything other than `false` with `language: 'none'`. (In the future, if the need arises, we can add support for linenums.)

kapunahelewong replied:
Hey @gkalpak, could you help me with this? I don't know how to do it.

gkalpak replied:
You can do something like this:

```diff
   private formatDisplayedCode() {
+    const linenums = this.getLinenums();
     const leftAlignedCode = leftAlign(this.code);
     this.setCodeHtml(leftAlignedCode); // start with unformatted code
     this.codeText = this.getCodeText(); // store the unformatted code as text (for copying)
 
     const skipPrettify = of(undefined);
     const prettifyCode = this.pretty
-        .formatCode(leftAlignedCode, this.language, this.getLinenums())
+        .formatCode(leftAlignedCode, this.language, linenums)
         .pipe(tap(formattedCode => this.setCodeHtml(formattedCode)));
+
+    if (linenums !== false && this.language === 'none') {
+      this.logger.warn(`Using 'linenums' with 'language: none' is currently not supported.`);
+    }
 
     // ...
   }
```

kapunahelewong replied:
Thank you so much!

gkalpak reviewed (changes requested):
BTW, I don't think we need to worry about e2e tests for this change.

For unit tests, ideally we should have tests that verify that `codeFormatted` emits (or remove the `codeFormatted` `@Output`, if it is not used anywhere), but it doesn't have to happen as part of this PR.

kapunahelewong commented on aio/src/app/custom-elements/code/code.component.spec.ts:
```diff
@@ -117,6 +117,14 @@ describe('CodeComponent', () => {
+      fixture.detectChanges();
+// ???
+      expect(getFormattedCode()).toBe(???);
+    });
```
Thank you so much!

mary-poppins commented:
You can preview f094524 at https://pr41335-f094524.ngbuilds.io/.

gkalpak commented on aio/src/app/custom-elements/code/code.component.ts:
```diff
@@ -111,12 +112,13 @@ export class CodeComponent implements OnChanges {
-    );
+        .pipe(tap(formattedCode => this.setCodeHtml(formattedCode)));
+
+    iif(() => this.language === 'none', skipPrettify, prettifyCode)
```
I probably proposed using `iif()` here, but looking at it again, since we are subscribing to the Observable ourselves, there is no need for `iif()` here. This could just be:

```suggestion
    (this.language === 'none' ? skipPrettify : prettifyCode)
```

kapunahelewong replied:
When I make this change, the next line causes issues. Should I also remove it (I think yes...?)? It's this one:

```
.subscribe(() => this.codeFormatted.emit(), () => { /* ignore failure to format */ });

```

gkalpak replied:
No, the following line should not be removed afaict. What kinds of issues are you getting?

gkalpak replied:
I pushed a fixup commit with a fix.

gkalpak commented on aio/src/app/custom-elements/code/code.component.ts:
```diff
@@ -111,12 +112,13 @@ export class CodeComponent implements OnChanges {
     this.codeText = this.getCodeText(); // store the unformatted code as text (for copying)
-
-    this.pretty
+    const skipPrettify = of(undefined);
```
```suggestion

    const skipPrettify = of(undefined);
```

Also, can you please log a warning when specifying linenums with `language === 'none'` (as mentioned in https://github.com/angular/angular/pull/41335#discussion_r600737153)?

gkalpak reviewed (changes requested):
The commit message should be `feat(docs-infra): ...` and contain a little more info on what it does (and what it is going to be used for).

mary-poppins commented:
You can preview be916cc at https://pr41335-be916cc.ngbuilds.io/.

kapunahelewong committed:
feat(docs-infra): add support for unformatted code-snippets

This commit adds support for skipping formatting in `<aio-code>`
elements (which are used by `<code-example>` and `<code-pane>` elements
under the hood) by specifying the `language` option as `'root'`.

This is useful for code-snippets that include plain text.

gkalpak committed:
fixup! feat(docs-infra): add support for unformatted code-snippets

gkalpak commented:
@googlebot I consent.

gkalpak commented:
@googlebot I consent.

gkalpak reviewed (approved):
FYI, I have pushed a fixup commit to fix the typing issue discussed in https://github.com/angular/angular/pull/41335#discussion_r623084473 and updated the commit message per https://github.com/angular/angular/pull/41335#pullrequestreview-627462880.

It should be good to go now 🚀
Thx, @kapunahelewong, for working on this ✨

mary-poppins commented:
You can preview f06542c at https://pr41335-f06542c.ngbuilds.io/.
